sdk/pinocchio/src/cpi.rs

```diff
@@ -102,48 +142,53 @@ pub fn slice_invoke_signed(
 ) -> ProgramResult {
-    if instruction.accounts.len() < account_infos.len() {
-        return Err(ProgramError::NotEnoughAccountKeys);
-    }
-
     if account_infos.len() > MAX_CPI_ACCOUNTS {
         return Err(ProgramError::InvalidArgument);
     }
 
     const UNINIT: MaybeUninit<Account> = MaybeUninit::<Account>::uninit();
     let mut accounts = [UNINIT; MAX_CPI_ACCOUNTS];
-    let mut len = 0;
 
-    for (account_info, account_meta) in account_infos.iter().zip(instruction.accounts.iter()) {
-        if account_info.key() != account_meta.pubkey {
-            return Err(ProgramError::InvalidArgument);
-        }
-
-        let state = if account_meta.is_writable {
-            BorrowState::Borrowed
-        } else {
-            BorrowState::MutablyBorrowed
-        };
-
-        if account_info.is_borrowed(state) {
-            return Err(ProgramError::AccountBorrowFailed);
-        }
-
-        // SAFETY: The number of accounts has been validated to be less than
-        // `MAX_CPI_ACCOUNTS`.
-        unsafe {
-            accounts
-                .get_unchecked_mut(len)
-                .write(Account::from(*account_info));
-        }
-
-        len += 1;
-    }
-    // SAFETY: The accounts have been validated.
+    account_infos
+        .iter()
+        .zip(instruction.accounts.iter())
+        .zip(accounts.iter_mut())
+        .try_for_each(|((account_info, account_meta), account)| {
+            // In order to check whether the borrow state is compatible
+            // with the invocation, we need to check that we have the
+            // correct account info and meta pair.
+            if account_info.key() != account_meta.pubkey {
+                return Err(ProgramError::InvalidArgument);
+            }
+
+            // Determines the borrow state that would be invalid according
+            // to their mutability on the instruction.
+            let invalid_state = if account_meta.is_writable {
+                // If the account is required to be writable, it cannot
+                //  be currently borrowed.
+                BorrowState::Borrowed
+            } else {
+                // If the account is required to be read-only, it cannot
+                // be currently mutably borrowed.
+                BorrowState::MutablyBorrowed
+            };
+
+            if account_info.is_borrowed(invalid_state) {
+                return Err(ProgramError::AccountBorrowFailed);
+            }
+
+            account.write(Account::from(*account_info));
+
+            Ok(())
+        })?;
+
+    // SAFETY: At this point it is guaranteed that all account infos are
+    // borrowable according to their mutability on the instruction, which
+    // in the worst case is more than what is needed.
     unsafe {
         invoke_signed_unchecked(
             instruction,
-            core::slice::from_raw_parts(accounts.as_ptr() as _, len),
+            core::slice::from_raw_parts(accounts.as_ptr() as _, account_infos.len()),
```

Comment:
Similarly, here, only the `min(instruction.accounts.len(), account_infos.len())` elements will actually be initialized.

Should we use a `min()` here?